src/libxtp/qmpair.cc

```diff
@@ -0,0 +1,214 @@
+/*
+ *            Copyright 2009-2018 The VOTCA Development Team
+ *                       (http://www.votca.org)
+ *
+ *      Licensed under the Apache License, Version 2.0 (the "License")
+ *
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *              http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+
+#include <votca/xtp/qmpair.h>
+#include <votca/xtp/segment.h>
+#include <votca/xtp/topology.h>
+
+using namespace votca::tools;
+
+namespace votca { namespace xtp {
+
+  QMPair::~QMPair() {
+    if (_ghost != NULL) {
+      delete _ghost; _ghost = NULL;
+    }
+  }
+
+
+  QMPair::QMPair(int id, Segment *seg1, Segment *seg2)
+    : std::pair<Segment*, Segment*>(seg1, seg2), _id(id),
+    _hasGhost(0),
+    _rate12_e(0), _rate21_e(0),
+    _rate12_h(0), _rate21_h(0),      
+    _has_e(false), _has_h(false),
+    _lambdaO_e(0), _lambdaO_h(0),
+    _Jeff2_e(0),   _Jeff2_h(0),
+    _rate12_s(0), _rate21_s(0),
+    _rate12_t(0), _rate21_t(0),
+    _has_s(false), _has_t(false),       
+    _lambdaO_s(0), _lambdaO_t(0),
+    _Jeff2_s(0),   _Jeff2_t(0),_pair_type( Hopping ) {
+
+      _top = seg1->getTopology();
+
+      vec r1 = seg1->getPos();
+      vec r2 = seg2->getPos();
+
+      _R = _top->PbShortestConnect(r1, r2); // => _R points from 1 to 2
+
+      // Check whether pair formed across periodic boundary
+      if ( abs(r2 - r1 - _R) > 1e-8 ) {
+
+        _ghost = new Segment(seg2);
+        _ghost->TranslateBy(r1 - r2 + _R);
+        _hasGhost = true;
+
+      }
+      else {
+        _ghost = NULL;
+      }
+    }
+
+
+  double QMPair::getLambdaO(int state) {
+    double result;
+    if (state ==-1) result=_lambdaO_e;
+    else if (state ==+1) result=_lambdaO_h;
+    else if (state ==+2) result=_lambdaO_s;
+    else if (state ==+3) result=_lambdaO_t;
+    else throw std::runtime_error(" ERROR CODE whx__01l1o__");
```

Comment:
I do not know what these error codes mean. Maybe a better Error message. We should store lambda in a vector and then get the number via an enum